astropy/io/fits/util.py

```diff
@@ -490,12 +502,38 @@ def _fileobj_normalize_mode(f):
             # This shouldn't happen?
             return None
 
+    # I've noticed that sometimes Python can produce modes like 'r+b' which I
+    # would consider kind of a bug--mode strings should be normalized. Let's
+    # normalize it for them:
     if '+' in mode:
         mode = mode.replace('+', '')
         mode += '+'
 
-    if _fileobj_is_append_mode(f) and 'a' not in mode:
-        mode = mode.replace('r', 'a').replace('w', 'a')
+    # There was (or still is) some format that loses or hides if it was
+    # opened in append mode... check it:
+
+    # TODO: This is probably superfluous the old corner cases where FILEIO
+    # objects didn't keep track of their append-flag are over (?)
+    # At least the bugfix was ported to python 2.7 and 3.3+:
+    # http://bugs.python.org/issue18876
```

Comment:
the warning is for cases I might have forgotten (the most common ones: pil, gzip, fileio, hdulist are covered in the tests) that still forget if they are opened in append-mode. These shouldn't exist anymore but better to have some kind of grace-period where users might get this warning.